src/fs_io.h

```diff
@@ -30,16 +30,25 @@ class dread_ctrl
     void *client_data;
 };
 
-// POD
 class dwrite_q
 {
+    MEMPROXY_CLASS(dwrite_q);
 public:
-    off_t file_offset;
-    char *buf;
-    size_t len;
-    size_t buf_offset;
-    dwrite_q *next;
-    FREE *free_func;
+    dwrite_q(const size_t wantCapacity) : dwrite_q(wantCapacity, nullptr, nullptr) {}
+    dwrite_q(const size_t hasLength, char *, FREE *);
```

Comment:
The `hasLength` was intentional to help clarify that oddity between the constructors usage of "size" parameter:
 * the first ctor is for creating an empty buffer with a given minimum capacity; 
 * the second ctor is setting an already filled buffer containing `hasLength` amount of content (thus minimum capacity of same). The `free_func` is (currently) required to have some hidden information about the real allocation size, so it is not passed in.

Hopefully this refactor will be able to make progress far enough to remove the weirdness.

Reply:
> The `hasLength` was intentional to help clarify that oddity between the constructors usage of "size" parameter:

That name did not help clarify. Please use one of the two suggested replacements instead (or propose a better replacement). Those suggestions also remove unwanted `const`.
 
> * the first ctor is for creating an empty buffer with a given minimum capacity;
> * the second ctor is setting an already filled buffer containing `hasLength` amount of content

Yes, I was aware of how the parameter is used when I wrote my change request.

> Hopefully this refactor will be able to make progress far enough to remove the weirdness.

I do not know what "this refactor" means in this context, but no PR changes have addressed this change request. Applying one of the suggested changes would address it.

Reply:
AFAICT you have requested one of two things:

 1) `dwrite_q(size_t bufCapacity, char *, FREE *);`
  * is a lie. This parameter is the **length** of valid content stored in the buffer. Callers passing the capacity will cause garbage outputs written when (capacity > Length).
 
 2) `dwrite_q(size_t, char *, FREE *);`
    * leaves developers guessing what the first parameter is for. Leading to false implications that it matches the other ctor where size **is** capacity. Back to the same bug as (1).

Reply:
>     2. `dwrite_q(size_t, char *, FREE *);`
>        
>  * leaves developers guessing what the first parameter is for. 

Yes, developers should look at the method implementation and, with some luck, will realize what this parameter means. This is far from ideal, of course, but better than any other name we have came up with so far -- I bet that any specific short name can be called a "lie" in this context because the true meaning is quite complex and cannot be fully captured by any short name.

More work is needed to properly fix this API problem. If I had the time, I would do that work, but I do not have the time. And I do not insist on that work done in this PR. Please pick the suggestion you dislike the least. None of them are worse than existing PR code.